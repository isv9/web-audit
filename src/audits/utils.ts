import { AuditResultLiveCollection, WebDocument } from '../web-audit'

export function getEmptyElementsLiveCollections (document: Pick<WebDocument, 'getEmptyElementsByTagName'>,
                                                 tagsNames: string[]): AuditResultLiveCollection[] {
  return tagsNames.reduce<AuditResultLiveCollection[]>((liveCollections, tag) => {
    const emptyElements = document.getEmptyElementsByTagName(tag)
    if (emptyElements.length > 0) {
      liveCollections.push({
        name: `Empty "${tag}" elements count in dom`,
        collection: emptyElements
      })
    }
    return liveCollections
  }, [])
}
