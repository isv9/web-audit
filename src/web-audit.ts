import { auditBlockSemantics, auditCommonSemantics, auditHeaderSemantics, auditTextSemantics } from './audits/semantic'
import { auditLinks } from './audits/links'

export type WebDocument = {
  getElementsByTagNameCount (tag: string): number;
  querySelectorAll (query: string): { length: number };
  getEmptyElementsByTagName (tag: string): { length: number }
};

type AuditSectionResult = {
  name: string;
  auditResults: AuditResult[];
};

export type AuditResult = {
  name?: string;
  tables: (AuditResultTable)[];
  liveCollections?: (AuditResultLiveCollection)[];
  logs?: string[];
  warnings?: string[];
  errors?: string[];
};

type AuditResultTable = { name?: string } & { [key: string]: string | number };
export type AuditResultLiveCollection = { name: string, collection: object };

export class WebAudit {

  private readonly webDocument: WebDocument

  constructor (webDocument: WebDocument) {
    this.webDocument = webDocument
  }

  audit () {
    const auditSemanticsResult = this.auditSemantics()
    const auditLinksResult = this.auditLinks()
    console.group('web audit')
    WebAudit.renderAuditSectionResult(auditSemanticsResult)
    WebAudit.renderAuditSectionResult(auditLinksResult)
    console.groupEnd()
  }

  private auditLinks (): AuditSectionResult {
    return {
      name: 'links',
      auditResults: [
        auditLinks(this.webDocument)]
    }
  }

  private auditSemantics (): AuditSectionResult {

    return {
      name: 'semantics',
      auditResults: [auditCommonSemantics(this.webDocument),
        auditHeaderSemantics(this.webDocument),
        auditTextSemantics(this.webDocument),
        auditBlockSemantics(this.webDocument)]
    }
  }

  static renderAuditSectionResult (auditSectionResult: AuditSectionResult) {
    const { name, auditResults } = auditSectionResult
    const details = {
      errorsCount: 0,
      warningsCount: 0
    }
    auditResults.forEach(auditResult => {
      const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
      const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
      details.errorsCount += errorsCount
      details.warningsCount += warningsCount
    })
    console.group(`${name}, (errors="${details.errorsCount}", warnings="${details.warningsCount}")`)
    auditResults.forEach(auditResult => WebAudit.renderAuditResult(auditResult))
    console.groupEnd()
  }

  static renderAuditResult (auditResult: AuditResult) {
    const { name, tables, errors = [], warnings = [], liveCollections = [], logs = [] } = auditResult
    const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
    const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
    console.group(`${name}, (errors="${errorsCount}", warnings="${warningsCount}")`)
    tables.forEach(({ name, nodes, ...tableDataSet }) => {
      console.log(name)
      console.table(tableDataSet)
    })
    liveCollections.forEach(({ name, collection }) => {
      console.log(name)
      console.log(collection)
    })
    logs.forEach(message => console.log(message))
    console.log('Summary')
    if (errorsCount > 0 || warningsCount > 0) {
      errors.forEach(message => console.error(message))
      warnings.forEach(message => console.warn(message))
    } else {
      console.log('It is OK')
    }

    console.groupEnd()
  }

  static getAuditResultMessagesCount (auditResult: AuditResult, key: keyof AuditResult): number {
    let property = auditResult[key]
    if (property === null || property === undefined) {
      property = []
    }
    if (Array.isArray(property)) {
      return property.length
    }
    throw new Error('Property type must be array')
  }
}

