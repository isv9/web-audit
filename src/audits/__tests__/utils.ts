import { WebDocument } from '../../web-audit';
import { getEmptyElementsLiveCollections } from '../utils';

describe('getEmptyElementsLiveCollections', () => {
  it('document does not have any empty elements', () => {
    const document: Pick<WebDocument, 'getEmptyElementsByTagName'> = {
      getEmptyElementsByTagName: jest.fn(() => ({ length: 0 })),
    };
    expect(getEmptyElementsLiveCollections(document, ['div', 'article'])).toMatchSnapshot();
  });
  it('document has some empty div elements', () => {
    const document: Pick<WebDocument, 'getEmptyElementsByTagName'> = {
      getEmptyElementsByTagName: jest.fn(tag => ({ length: tag === 'div' ? 5 : 0 })),
    };
    expect(getEmptyElementsLiveCollections(document, ['div', 'article'])).toMatchSnapshot();
  });
});
