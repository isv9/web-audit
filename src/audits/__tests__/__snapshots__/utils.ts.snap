// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEmptyElementsLiveCollections document does not have any empty elements 1`] = `Array []`;

exports[`getEmptyElementsLiveCollections document has some empty div elements 1`] = `
Array [
  Object {
    "collection": Object {
      "length": 5,
    },
    "name": "Empty \\"div\\" elements count in dom",
  },
]
`;
